parser/scanner.go

```diff
@@ -7,75 +7,99 @@ import (
+	sliceStart  int    // the start of the slice visible to the scanner
+	sliceLength int    // the length of the slice visible to the scanner
 }
 
-func NewScanner(src string) *Scanner {
-	return &Scanner{src: src, slice: src, offset: 0}
+type source interface {
+	length() int                // the length of the entire source string
+	slice(i, length int) string // the string of the given slice
+	filename() string           // the name of the file from which the source is derived (or empty if none)
 }
 
-func NewScannerAt(src string, offset, size int) *Scanner {
-	return &Scanner{src: src, slice: src[offset : offset+size], offset: offset}
+type stringSource struct {
+	origin *string // the entire source string
+	f      string  // the source filename
 }
 
-func NewBareScanner(offset int, slice string) *Scanner {
-	return &Scanner{slice: slice, offset: offset}
+func NewScanner(str string) *Scanner {
+	return &Scanner{stringSource{origin: &str}, 0, len(str)}
 }
 
-func (r Scanner) String() string {
-	return r.slice
+func NewScannerWithFilename(str, filename string) *Scanner {
+	return &Scanner{stringSource{&str, filename}, 0, len(str)}
 }
 
-func (r Scanner) StripSource() Scanner {
-	r.src = ""
-	return r
+func NewScannerAt(str string, offset, size int) *Scanner {
+	return &Scanner{stringSource{origin: &str}, offset, size}
 }
 
-func (r Scanner) Format(state fmt.State, c rune) {
+// - Scanner
+
+// The name of the file from which the source is derived (or empty if none).
+func (s Scanner) Filename() string {
+	return s.src.filename()
+}
+
+func (s Scanner) String() string {
+	if s.src == nil {
+		return ""
+	}
+	return s.slice()
+}
+
+func (s Scanner) Format(state fmt.State, c rune) {
 	if c == 'q' {
-		fmt.Fprintf(state, "%q", r.String())
+		_, _ = fmt.Fprintf(state, "%q", s.slice())
 	} else {
-		state.Write([]byte(r.String()))
+		_, _ = state.Write([]byte(s.slice()))
 	}
 }
 
-func (r Scanner) Context() string {
+func (s Scanner) Context() string {
+	end := s.sliceStart + s.sliceLength
 	return fmt.Sprintf("%s\033[1;31m%s\033[0m%s",
-		r.src[:r.offset],
-		r.slice,
-		r.src[r.offset+len(r.slice):],
+		s.src.slice(0, s.sliceStart),
+		s.slice(),
+		s.src.slice(end, s.src.length()-end),
 	)
 }
 
-func (r Scanner) Offset() int {
-	return r.offset
+// The position of the start of the scanner within the original source.
+func (s Scanner) Offset() int {
+	return s.sliceStart
 }
 
-func (r Scanner) Slice(a, b int) *Scanner {
-	return &Scanner{
-		src:    r.src,
-		slice:  r.slice[a:b],
-		offset: r.offset + a,
-	}
+// The 1-indexed line and column number of the start of the scanner within the original source.
+func (s Scanner) Position() (int, int) {
+	return lineColumn(s.src.slice(0, s.sliceStart), s.sliceStart)
+}
+
+// The slice that is visible to the scanner
+func (s Scanner) slice() string {
+	return s.src.slice(s.sliceStart, s.sliceLength)
+}
+
+func (s Scanner) Slice(a, b int) *Scanner {
+	return &Scanner{s.src, s.sliceStart + a, b - a}
```

Comment:
shouldn't this be `b - a + 1`

Reply:
No. A slice from 0 to 0 should have a length of 0, not 1.